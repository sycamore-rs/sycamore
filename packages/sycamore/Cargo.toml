[package]
name = "sycamore"
categories = ["gui", "wasm", "web-programming"]
description = "A library for building reactive web apps in Rust and WebAssembly"
edition = "2021"
homepage = "https://github.com/sycamore-rs/sycamore"
keywords = ["wasm", "gui", "reactive"]
license = "MIT"
readme = "../../README.md"
repository = "https://github.com/sycamore-rs/sycamore"
version = "0.8.0"

[dependencies]
ahash = "0.7.6"
futures = { version = "0.3.21", optional = true }
html-escape = { version = "0.2.9", optional = true }
indexmap = { version = "1.8.0", features = ["std"] }
js-sys = { version = "0.3.56", optional = true }
once_cell = { version = "1.10.0", optional = true }
paste = "1.0.6"
sycamore-core = { path = "../sycamore-core", version = "0.8.0" }
sycamore-futures = { path = "../sycamore-futures", version = "0.8.0", optional = true }
sycamore-macro = { path = "../sycamore-macro", version = "0.8.0" }
sycamore-reactive = { path = "../sycamore-reactive", version = "0.8.0" }
sycamore-web = { path = "../sycamore-web", version = "0.8.0", optional = true }
wasm-bindgen = { version = "0.2.79", optional = true }
wasm-bindgen-futures = { version = "0.4.29", optional = true }

[dependencies.web-sys]
features = [
    "console",
    "Comment",
    "Document",
    "DocumentFragment",
    "DomTokenList",
    "Element",
    "Event",
    "HtmlElement",
    "HtmlInputElement",
    "HtmlCollection",
    "Node",
    "NodeList",
    "Text",
    "Window",
]
version = "0.3.56"
optional = true

[dev-dependencies]
criterion = "0.3.5"
expect-test = "1.2.2"
tokio = { version = "1.17.0", features = ["macros", "rt"] }
wasm-bindgen-test = "0.3.29"

[features]
default = ["web", "wasm-bindgen-interning"]
hydrate = [
    "web",
    "sycamore-macro/hydrate",
    "sycamore-core/hydrate",
    "sycamore-web/hydrate",
]
ssr = [
    "web",
    "html-escape",
    "once_cell",
    "hydrate",
    "sycamore-macro/ssr",
    "sycamore-web/ssr",
]
suspense = ["futures", "wasm-bindgen-futures", "sycamore-futures"]
serde = ["sycamore-reactive/serde"]
wasm-bindgen-interning = ["web", "wasm-bindgen/enable-interning"]
web = ["wasm-bindgen", "web-sys", "js-sys", "sycamore-web"]

[[bench]]
harness = false
name = "reactivity"

[[bench]]
harness = false
name = "ssr"
required-features = ["ssr"]

[[test]]
name = "ssr"
required-features = ["ssr"]

[package.metadata.docs.rs]
all-features = true
default-target = "wasm32-unknown-unknown"
