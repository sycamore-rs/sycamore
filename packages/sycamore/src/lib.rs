//! # Sycamore API Documentation
//!
//! Sycamore is a VDOM-less web library with fine-grained reactivity.
//!
//! This is the API docs for sycamore. If you are looking for the usage docs, checkout the
//! [Sycamore Book](https://sycamore-rs.netlify.app/docs/getting_started/installation).
//!
//! ## Feature Flags
//! - `dom` (_default_) - Enables rendering templates to DOM nodes. Only useful on
//!   `wasm32-unknown-unknown` target.
//! - `experimental-builder-agnostic` - Enables the agnostic backend builder API.
//! - `experimental-builder-html` - Enables the HTML specific backend builder API. Also enables
//!   `experimental-builder-agnostic`.
//! - `experimental-hydrate` - Enables client-side hydration support.
//! - `futures` - Enables wrappers around `wasm-bindgen-futures` to make it easier to extend a
//!   reactive scope into an `async` function.
//! - `ssr` - Enables rendering templates to static strings (useful for Server Side Rendering /
//!   Pre-rendering).
//! - `serde` - Enables serializing and deserializing `Signal`s and other wrapper types using
//!   `serde`.
//! - `wasm-bindgen-interning` (_default_) - Enables interning for `wasm-bindgen` strings. This
//!   improves performance at a slight cost in binary size. If you want to minimize the size of the
//!   result `.wasm` binary, you might want to disable this.

#![warn(clippy::clone_on_ref_ptr)]
#![warn(clippy::rc_buffer)]
#![warn(clippy::semicolon_if_nothing_returned)]
#![warn(rust_2018_idioms)]
#![deny(clippy::trait_duplication_in_bounds)]
#![deny(clippy::type_repetition_in_bounds)]

// Alias self to `sycamore` to make it possible to use proc-macros within the `sycamore` crate.
#[allow(unused_extern_crates)] // False positive
extern crate self as sycamore;

#[cfg(feature = "experimental-builder-agnostic")]
pub mod builder;
pub mod component;
pub mod flow;
#[cfg(feature = "futures")]
pub mod futures;
pub mod generic_node;
pub mod motion;
pub mod noderef;
pub mod utils;
pub mod view;

pub mod reactive {
    pub use sycamore_reactive::*;
}

#[cfg(feature = "ssr")]
pub use crate::generic_node::render_to_string;
#[cfg(all(feature = "dom", feature = "experimental-hydrate"))]
pub use generic_node::{hydrate, hydrate_get_scope, hydrate_to};
#[cfg(feature = "dom")]
pub use generic_node::{render, render_get_scope, render_to};

pub use sycamore_macro::*;

/// The sycamore prelude.
pub mod prelude {
    #[cfg(feature = "dom")]
    pub use crate::generic_node::DomNode;
    #[cfg(all(feature = "dom", feature = "experimental-hydrate"))]
    pub use crate::generic_node::HydrateNode;
    #[cfg(feature = "ssr")]
    pub use crate::generic_node::SsrNode;

    pub use crate::flow::*;
    pub use crate::generic_node::{GenericNode, Html};
    pub use crate::noderef::{NodeRef, ScopeCreateNodeRef};
    pub use crate::reactive::*;
    pub use crate::view::View;

    pub use sycamore_macro::*;
}

/// Re-exports for use by `sycamore-macro`. Not intended for use by end-users.
#[doc(hidden)]
pub mod rt {
    pub use js_sys::Reflect;
    pub use wasm_bindgen::{intern, JsCast, JsValue};
    pub use web_sys::Event;
}
