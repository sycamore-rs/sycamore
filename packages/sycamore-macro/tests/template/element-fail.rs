use sycamore::prelude::*;

fn compile_fail<G: GenericNode>() {
    let _: TemplateResult<G> = template! { p.my-class#id };

    let _: TemplateResult<G> = template! { button(disabled) };
    let _: TemplateResult<G> = template! { button(on:click) };
    let _: TemplateResult<G> = template! { button(unknown:directive="123") };

    let _: TemplateResult<G> = template! { button(a.b.c="123") };
}

fn main() {}
