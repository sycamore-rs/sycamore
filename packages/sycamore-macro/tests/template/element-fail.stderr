error: expected a valid HTML node
 --> $DIR/element-fail.rs:4:39
  |
4 |     let _: Template<G> = template! { p.my-class#id };
  |                                       ^

error: expected `=`
 --> $DIR/element-fail.rs:6:44
  |
6 |     let _: Template<G> = template! { button(disabled) };
  |                                            ^^^^^^^^^^

error: expected `=`
 --> $DIR/element-fail.rs:7:44
  |
7 |     let _: Template<G> = template! { button(on:click) };
  |                                            ^^^^^^^^^^

error: unknown directive `unknown`
 --> $DIR/element-fail.rs:8:45
  |
8 |     let _: Template<G> = template! { button(unknown:directive="123") };
  |                                             ^^^^^^^

error: expected `=`
  --> $DIR/element-fail.rs:10:46
   |
10 |     let _: Template<G> = template! { button(a.b.c="123") };
   |                                              ^

error: children and inner html cannot be both set
  --> $DIR/element-fail.rs:12:26
   |
12 |       let _: Template<G> = template! {
   |  __________________________^
13 | |         p(dangerously_set_inner_html="<span>Test</span>") {
14 | |             "Error"
15 | |         }
16 | |     };
   | |_____^
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
